src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs

```diff
@@ -833,7 +833,8 @@ private static bool IsW3CId(string id)
             // The version (00-fe) is used to indicate that this is a WC3 ID.
             return id.Length == 55 &&
                    ('0' <= id[0] && id[0] <= '9' || 'a' <= id[0] && id[0] <= 'f') &&
-                   ('0' <= id[1] && id[1] <= '9' || 'a' <= id[1] && id[1] <= 'e');
+                   ('0' <= id[1] && id[1] <= '9' || 'a' <= id[1] && id[1] <= 'f') &&
+                   (id[0] != 'f' || id[1] != 'f');
```

Comment:
Since this project is compiling HexConverter.cs from common path, this can be simplified to:

```suggestion
            return id.Length == 55 &&
				     16 * HexConverter.FromChar(id[0]) + HexConverter.FromChar(id[1]) < 255;
```
also cuts the codegen by 40%.

Reply:
any perf concern with this suggestion?

Reply:
For codegen, I was unintentionally using Debug configuration (which was showing 40% reduction in number of instructions). In release mode, HexConverter.FromChar (which uses array lookup) regresses in number of instructions, and also regresses in perf by ratio 1.23. HexConverter.FromLowerChar regresses even worse by ration 1.33, as it does not use array, but instead has branch with arithmetic operations.

@xiang17, you can revert it to use  your original method. :)

Reply:
For codegen, I was unintentionally using Debug configuration (which was showing 40% reduction in number of instructions). In release mode, HexConverter.FromChar (which uses array lookup) regresses in number of instructions, and also regresses in perf by ratio 1.23. HexConverter.FromLowerChar regresses even worse by ratio 1.33, as it does not use array, but instead has branch with arithmetic operations.

@xiang17, you can revert it to use  your original method. :)

Reply:
|                Method |                   id |     Mean |     Error |    StdDev | Ratio | RatioSD |
|---------------------- |--------------------- |---------:|----------:|----------:|------:|--------:|
|               IsW3CId | ffxxx(...)xxxxx [55] | 2.214 ns | 0.0261 ns | 0.0218 ns |  1.00 |    0.00 |
|      IsW3CId_FromChar | ffxxx(...)xxxxx [55] | 2.688 ns | 0.0841 ns | 0.1258 ns |  1.23 |    0.03 |
| IsW3CId_FromLowerChar | ffxxx(...)xxxxx [55] | 2.950 ns | 0.0341 ns | 0.0302 ns |  1.33 |    0.02 |

Reply:
I didn't refresh and didn't see tarekgh's comment when I was making the commit "Use bitwise operator for both operation better readability". I've reverted it back to my original method now.

By the way, I'm new to this repository and curious how codegen performance is measured and/or monitored. Could you point me where are the tests from which you got the table? Or did you just wrote some tests for evaluating codegen performance?

Reply:
@xiang17, for codegen, we need to first perform a checked build `./build.sh -rc Checked -c Release`, then use the compiled runtime to run your app with `COMPlus_JitDisasm` set to the name or names of methods you want to disassemble. e.g. `COMPlus_JitDisasm=* dotnet myapp.dll` to view disassembly of all the methods or `COMPlus_JitDisasm=MyMethod1,MyMethod2` to view targeted assembly code. We can also use https://sharplab.io/, which is currently using .NET 5, so it might be missing optimizations/regressions in master branch since .NET 5 release (but it gives us a quick/rough ideas).

for performance measurement, I used benchmarkdotnet template https://www.nuget.org/packages/BenchmarkDotNet.Templates/:
```sh
$ dotnet new --install BenchmarkDotNet.Templates::0.12.1
$ dotnet new benchmark --console-app -n gh-46492
$ cd gh-46492
# replace some stuff since the template is outdated https://github.com/dotnet/BenchmarkDotNet/issues/1512
$ sed -i '' 's/netcoreapp3.0/net5.0/' gh-46492.csproj
$ sed -i '' 's/0.12.0/0.12.1/' gh-46492.csproj
$ curl -sSLO https://raw.githubusercontent.com/dotnet/runtime/e87e32f50caa4122cf689eebb3810989dee6e125/src/libraries/Common/src/System/HexConverter.cs
$ cat > Benchmarks.cs <<EOF
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Runtime.CompilerServices;
using BenchmarkDotNet;
using BenchmarkDotNet.Attributes;

namespace gh_46492
{
    public class Benchmarks
    {
        // public field
        [ParamsSource(nameof(ValuesForB))]
        public string id;

        public static IEnumerable<string> ValuesForB() => new[]
        {
            "00xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx",
            "fexxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx",
            "ffxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
        };

        [Benchmark(Baseline = true)]
        public bool IsW3CId()
        {
            return id.Length == 55 &&
                   ('0' <= id[0] && id[0] <= '9' || 'a' <= id[0] && id[0] <= 'f') &&
                   ('0' <= id[1] && id[1] <= '9' || 'a' <= id[1] && id[1] <= 'f') &&
                   (id[0] != 'f' || id[1] != 'f');
        }

        [Benchmark]
        public bool IsW3CId_FromChar()
        {
            return id.Length == 55 &&
                   HexConverter.IsHexChar(id[0]) && HexConverter.IsHexChar(id[1]) &&
                   16 * HexConverter.FromChar(id[0]) + HexConverter.FromChar(id[1]) < 255;
        }

        [Benchmark]
        public bool IsW3CId_FromLowerChar()
        {
            return id.Length == 55 &&
                   HexConverter.IsHexChar(id[0]) && HexConverter.IsHexChar(id[1]) &&
                   ((HexConverter.FromLowerChar(id[0]) << 4) + HexConverter.FromLowerChar(id[1])) < 255;
        }
    }
}
EOF
$ dotnet run -c Release
```

Reply:
That's cool! Thanks.